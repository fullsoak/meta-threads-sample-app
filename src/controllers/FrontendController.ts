import { Context, Controller, Get, ssr } from "fullsoak";
import { makeHat } from "fullsoak/batteries";
import { Home } from "../components/Home/index.tsx";
import { isLoggedIn } from "../utils/db.ts";
import { getThreadProfile } from "../services/getThreadsProfile.ts";

const defaultThreadsAppScope =
  "threads_basic,threads_manage_mentions,threads_content_publish";

const getThreadsLoginUrl = (): string => {
  const redirectUri = encodeURIComponent(
    Deno.env.get("THREADS_AUTH_CODE_REDIRECT_URI") || "",
  );
  return "https://threads.net/oauth/authorize" +
    `?client_id=${Deno.env.get("THREADS_APP_ID")}` +
    `&redirect_uri=${redirectUri}` +
    `&scope=${
      Deno.env.get("THREADS_APP_AUTH_SCOPE") || defaultThreadsAppScope
    }` +
    "&response_type=code";
};

const headContent = makeHat({
  title: "Meta Threads Sample App :: FullSoak",
  links: [{
    rel: "stylesheet",
    href:
      "https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.20.1/cdn/themes/light.css",
  }, {
    rel: "stylesheet",
    href: "https://fonts.googleapis.com/css2?family=DM%20Mono",
  }],
  scripts: [{
    type: "module",
    src:
      "https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.20.1/cdn/shoelace-autoloader.js",
  }],
});

@Controller()
export class FrontendController {
  @Get("/")
  async home(ctx: Context) {
    const userId = await ctx.cookies.get("auth");
    const loggedIn = isLoggedIn(userId);
    return ssr(Home, {
      loggedIn,
      threadsLoginUrl: getThreadsLoginUrl(),
      threadsProfile: loggedIn ? await getThreadProfile(userId) : undefined,
    }, { headContent });
  }
}
