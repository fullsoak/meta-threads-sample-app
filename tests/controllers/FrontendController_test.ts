import { superoak } from "superoak";
import { assertSnapshot } from "jsr:@std/testing/snapshot";
import { FrontendController } from "../../src/controllers/FrontendController.ts";
import { useFullSoak } from "fullsoak/testing";

const app = await useFullSoak({ controllers: [FrontendController] });

Deno.test("GET / should response with expected HTML content", async (t) => {
  const request = await superoak(app);
  const resp = await request.get("/").expect(200);
  await assertSnapshot(t, resp.text);
});
