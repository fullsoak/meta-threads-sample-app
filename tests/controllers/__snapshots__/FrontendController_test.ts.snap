export const snapshot = {};

snapshot[`GET / should response with expected HTML content 1`] = `
'<!doctype html><html><head><meta charset="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0"/><script type="importmap">{"imports":{"preact":"https://esm.sh/preact@10.26.4","preact/hooks":"https://esm.sh/preact@10.26.4/hooks","preact/debug":"https://esm.sh/preact@10.26.4/debug","preact/devtools":"https://esm.sh/preact@10.26.4/devtools","react":"https://esm.sh/preact@10.26.4","react/jsx-runtime":"https://esm.sh/preact@10.26.4/jsx-runtime","preact/jsx-runtime":"https://esm.sh/preact@10.26.4/jsx-runtime","react-dom":"https://esm.sh/preact@10.26.4/compat/","react-dom/*":"https://esm.sh/preact@10.26.4/compat/*","htm/preact":"https://esm.sh/htm@3.1.1/preact?external=preact","preact-iso":"https://esm.sh/preact-iso@^2.9.1?external=preact","preact-iso/prerender":"/fullsoak","@preact/signals":"https://esm.sh/@preact/signals","fullsoak/preact-iso":"https://esm.sh/preact-iso@^2.9.1?external=preact","@fullsoak/fullsoak/preact-iso":"https://esm.sh/preact-iso@^2.9.1?external=preact","fullsoak":"/fullsoak","@fullsoak/fullsoak":"/fullsoak"}}</script><title>Meta Threads Sample App :: FullSoak</title><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.20.1/cdn/themes/light.css"/><link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=DM%20Mono"/><script type="module" src="https://cdn.jsdelivr.net/npm/@shoelace-style/shoelace@2.20.1/cdn/shoelace-autoloader.js"></script><link rel="modulepreload" href="/components/Home/mount" as="script" type="text/javascript"/><link rel="modulepreload" href="/components/Home/index.tsx" as="script" type="text/javascript"/><script type="text/javascript"></script><style>:root{--background-color:#fff5f6;--button-color:#000000;--button-hover-color:#333333}*{margin:0;padding:0;box-sizing:border-box}body{background-color:var(--background-color);min-height:100vh;display:flex;justify-content:center;align-items:center;font-size:16px;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif}.title{color:#901811}.title+.subtitle{margin-top:1.5em;margin-bottom:1.5em}.background-ribbon{position:fixed;top:0;left:0;width:100%;height:100%;background-image:url("ribbon.svg");background-size:cover;background-position:center;opacity:.1;pointer-events:none;z-index:0}.container{min-height:100vh;position:relative;padding:1.5em;z-index:1}.landing-box{min-height:100vh;align-content:center}.container ul{list-style:none;margin-bottom:1.5em}.login-button{display:inline-block;background-color:var(--button-color);color:#fff;border:none;padding:15px 32px;font-size:16px;border-radius:8px;cursor:pointer;transition:background-color .3s ease;font-weight:600;text-decoration:none;margin-top:.5em}.plate{display:flex;flex-direction:row;gap:1em}@media screen and (max-width:375px){.plate{flex-direction:column}}.plate-toolbar{display:flex;flex-direction:column;gap:1em}@media screen and (min-width:376px){.plate-toolbar{width:500px;max-width:500px;min-width:320px}}.mention-card:hover{--sl-panel-background-color:var(--sl-color-sky-100)}.mention-box{display:flex;flex-direction:column;gap:.25em}.mention-box sl-textarea{margin-top:1em;margin-bottom:.25em}.card-comment-box{display:flex;flex-direction:column}.card-comment-box sl-textarea{margin-bottom:.5em}.login-button:hover{background-color:var(--button-hover-color)}section.main div+div{margin-top:1.5em}a.clean{text-decoration:none;color:inherit}.external-btn{margin-left:.25em}.external-btn sl-icon{font-size:.75em}</style></head><body><main id="fullsoak-htmlshell-main"><div class="background-ribbon"></div><div class="container"><div class="landing-box"><h1 class="title">Red Threads :: Thread Roller</h1><a class="login-button" href="https://threads.net/oauth/authorize?client_id=undefined&amp;redirect_uri=&amp;scope=threads_basic,threads_manage_mentions,threads_content_publish&amp;response_type=code">Log In with Threads Account</a></div></div><style>
  .card-overview {
    max-width: 350px;
  }

  .card-overview small {
    color: var(--sl-color-neutral-500);
  }

  .card-overview [slot="footer"] {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
</style></main><script type="text/javascript">window.preloadedProps = {"loggedIn":false,"threadsLoginUrl":"https://threads.net/oauth/authorize?client_id=undefined&redirect_uri=&scope=threads_basic,threads_manage_mentions,threads_content_publish&response_type=code"}</script><script defer type="module" src="/components/Home/mount"></script></body></html>'
`;

snapshot[`GET /unknown-route should return 404 1`] = `""`;
